tedge/src/privsep.rs

```diff
@@ -0,0 +1,92 @@
+use std::io::{BufRead, BufReader, Read, Write};
+
+#[derive(Debug)]
+pub enum PrivilegedCommand {
+    Command1,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum PrivilegedCommandError {
+    #[error("Insufficient privileges: Try running with `sudo`")]
+    InsufficientPrivileges,
+
+    #[error("Command execution failed")]
+    CommandFailed,
+}
+
+pub trait PrivilegedCommandExecutor {
+    fn execute(&mut self, command: PrivilegedCommand) -> Result<(), PrivilegedCommandError>;
+}
+
+/// We use this executor when not running as `root`. It will always fail with an error.
+pub struct UnprivilegedDummyCommandExecutor;
+
+impl PrivilegedCommandExecutor for UnprivilegedDummyCommandExecutor {
+    fn execute(&mut self, command: PrivilegedCommand) -> Result<(), PrivilegedCommandError> {
```

Comment:
I don't think it can be as simple as this because when not running as root, we still need to do these commands, we just do slightly different things - for example writing a config file in a different place or starting a process rather than using systemd.